Third-party-libraries.md

```diff
@@ -0,0 +1,255 @@
+
+  The object must also specify one of the following keys:
+
+  * `targetDir`: The base name to which the expanded file or folder should be renamed.
+  * `targetDirPrefix`: The prefix of the base name to which the expanded file or folder should be renamed. The prefix concatenated with the version string should produce the full base name.
+
+* Collections of files. The following keys are required in the object describing this dependency:
+
+  * `version`: The dependency's version string.
+  * `url`: The prefix of the URL to the files.
+  * `downloadFormat`: Must be set to `files`.
+  * `files`: List of the files to download. Each file will be downloaded from the URL formed by joining `url` with the item in `files`, using a slash (`/`) as a delimiter. Files are not renamed after they are downloaded.
+  * `targetDirPrefix`: A string which, when suffixed with `version`, yields the base name of the folder to which each file will be downloaded.
+
+* Tar archives of files. These work just like zip archives, except that there are no optional keys. Instead of the `rootDir`, `rootDirPrefix`, `targetDir`, and `targetDirPrefix` keys, these two keys are required:
+
+  * `tarRootDirPrefix`: Same as `rootDirPrefix` for zip files.
+  * `targetDirPrefix`: Same as `targetDirPrefix` for zip files.
+
+New dependencies should not be added to `manifest.json`, as this method of installing dependencies is deprecated. Instead, you should use the [node modules](#node-modules) method.
+
+#### Redis and Elasticsearch
+
+We download and install the Redis CLI and Elasticsearch development server. We pin the versions of these dependencies that we install, but we don't automatically upgrade old versions. Specifically, we try to execute the Redis and Elasticsearch binaries. If the binaries execute successfully, then we don't install anything.
+
+### `install_backend_python_libs.py`
+
+This script installs the Python dependencies we need in production to `third_party/python_libs`. We define these dependencies using `requirements.in`, which lists the libraries we depend on directly. Then `install_backend_python_libs.py` runs `scripts.regenerate_requirements` to generate `requirements.txt`, which lists those direct dependencies, plus all the packages that our direct dependencies need, and so on. Both `requirements.in` and `requirements.txt` specify versions, so `requirements.txt` is analogous to `yarn.lock` in that it pins all the versions of all the Python packages we use in production.
+
+## Upgrade dependencies
+
+### Upgrade backend libraries
+
+Before upgrading any backend dependencies, you should check for any breaking changes between the currently installed version and the new version you want to upgrade the library to. You should:
+
+* Check the library's changelog for breaking changes.
+* Test that after you install the upgraded version (see below), everything works correctly.
```

Comment:
Probably should add comma before 'after' since the "after ... (see below)" part is a secondary clause. Or put "everything works correctly" before "after".

Maybe also add **This is very important.** after this sentence.

Reply:
done